File: vuln-sandbox/go/cwe-327/weak-crypto-safe.go
// SAFE: go-des-cipher-usage — AES-256-GCM used instead of weak DES cipher
// Rule: go-des-cipher-usage | CWE-327 | Expected: TrueNegative

package main

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"fmt"
	"io"
)

func encrypt(plaintext []byte, key []byte) ([]byte, error) {
	// SAFE: AES-256-GCM is a strong authenticated encryption algorithm; DES is not used
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}

	nonce := make([]byte, gcm.NonceSize())
	if _, err = io.ReadFull(rand.Reader, nonce); err != nil {
		return nil, err
	}

	return gcm.Seal(nonce, nonce, plaintext, nil), nil
}

func main() {
	key := make([]byte, 32) // 256-bit key
	rand.Read(key)

	ciphertext, err := encrypt([]byte("sensitive data"), key)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}
	fmt.Printf("Encrypted %d bytes\n", len(ciphertext))
}

File: vuln-sandbox/rust/cwe-327/weak_crypto_safe.rs
// SAFE: rust-generic-format-sql — AES-256-GCM used instead of weak hash for data protection
// Rule: rust-generic-format-sql | CWE-327 | Expected: TrueNegative

use aes_gcm::{
    aead::{Aead, AeadCore, KeyInit, OsRng},
    Aes256Gcm, Key, Nonce,
};

fn encrypt(plaintext: &[u8], key: &Key<Aes256Gcm>) -> Result<Vec<u8>, aes_gcm::Error> {
    // SAFE: AES-256-GCM is a strong authenticated encryption algorithm
    let cipher = Aes256Gcm::new(key);
    let nonce = Aes256Gcm::generate_nonce(&mut OsRng);

    let ciphertext = cipher.encrypt(&nonce, plaintext)?;

    // Prepend nonce to ciphertext for storage
    let mut result = nonce.to_vec();
    result.extend_from_slice(&ciphertext);
    Ok(result)
}

fn main() {
    let key = Aes256Gcm::generate_key(OsRng);
    match encrypt(b"sensitive data", &key) {
        Ok(ciphertext) => println!("Encrypted {} bytes", ciphertext.len()),
        Err(e) => eprintln!("Error: {}", e),
    }
}
